---
title: "F<sub>ST</sub>"
---

<!-- add additional code/examples here, Ryan will add additional code and custom R script here-->

# Calculating Pairwise *F<sub>ST</sub>*

1. Obtain scripts: `SAF.sh` & `pwfst.sh`

2. Create a bamlist with individuals from only the "population" you want to test

3. Generate a simple allele frequencies file (`.saf`) for every "population" bamlist you want to compare, requires bamlist, outfile name, and your ancestral/alignment file (i.e., *`___300.fasta`*)
    - `sbatch -t 24:00:00 - p high SAF.sh bamlist outfile_name ancestral_file`
  
4. For pairwise comparisons between "populations", use two `____.saf.index` files you just generated:
    - `sbatch -t 24:00:00 -p high pwfst.sh pop1 pop2`  where `pop1` is the filename (no extension) of the 1st `___.saf.index` and `pop2` is the 2nd
    - find the output files: `ls *finalfstout` to view. The adjusted **F<sub>ST</sub>** is the second value.
