# Miller Lab RADSeq & RAPTURE Pipeline

RADSeq & RAPTURE code and steps. This site is an [Rmarkdown website](http://rmarkdown.rstudio.com/rmarkdown_websites.html), and can be built using: `rmarkdown::render_site()`.

See the [`scripts`](https://github.com/ryanpeek/radseq/tree/master/scripts) folder for raw `R` and `bash` scripts.

TO DO:

 - make filename consistent (and neutral across instructions)
 - make paths relative and consistent (where possible) globally
 - add colony tab
 - angsd tab?, change angsd to SNP/Genotype calling
 - PIPELINES (examples)
 - Alignment (De Novo, from Reference Genome, Merging)
 - Filtering/Data Quality: subsampling
 - Popgen Analysis: Tools (angsd & samtools), short explanation and breaking down the main pieces of an angsd call, bulleted list of what diff analysis are done with each tool...then actual analysis.
 - Post-Analysis (Archiving Data): NCBI