<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Downloading Illumina Sequence Data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />




<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 61px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 66px;
  margin-top: -66px;
}

.section h2 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h3 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h4 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h5 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h6 {
  padding-top: 66px;
  margin-top: -66px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Background
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="background_bash.html">Using bash</a>
    </li>
    <li>
      <a href="sftp.html">Using sftp</a>
    </li>
    <li>
      <a href="resources.html">Additional Coding Resources</a>
    </li>
  </ul>
</li>
<li>
  <a href="downloading_illumina.html">Downloading Data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="de_novo_assembly.html">de Novo Assembly</a>
    </li>
    <li>
      <a href="reference.html">Reference Assembly</a>
    </li>
    <li>
      <a href="merging.html">Merging Reads</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Filtering/Data Quality
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="subsampling.html">Subsampling</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    PopGen Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="popgen_tools.html">Tools</a>
    </li>
    <li>
      <a href="pca.html">PCA</a>
    </li>
    <li>
      <a href="admixture.html">Admixture</a>
    </li>
    <li>
      <a href="FST.html">F_st</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Piplelines
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline_colony.html">Colony</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Post-Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="upload_ncbi.html">Uploading to NCBI</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="contact.html">Contact</a>
</li>
<li>
  <a href="https://github.com/ryanpeek/radseq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Downloading Illumina Sequence Data</h1>

</div>


<div id="general-genomic-information" class="section level2">
<h2>General Genomic Information</h2>
<p>You’ll need to know which cutter you are using, so you can pull/use the appropriate barcode and splitting scripts. Basically, each sample/file will end up labeled with a Biotin <strong>Plate</strong> barcode, and an individual <strong>Well</strong> barcode.</p>
<p>These will look something like (for <em>Sbf1</em>):</p>
<ul>
<li><em>PlateBarcode</em>-CC_CCTGCA_GG_<em>FRAGMENT</em></li>
</ul>
<div id="cutters" class="section level3">
<h3>Cutters</h3>
<p>The two cutters we primarily use are:</p>
<ul>
<li>Six cutter <code>(5' -  CTGCA_G - 3')</code> <em>Pst1</em></li>
<li>Eight cutter <code>(5' - CCTGCA_GG - 3')</code> <em>Sbf1</em></li>
</ul>
<div id="q-phred-quality-scores" class="section level4">
<h4>Q Phred Quality scores</h4>
<p>Initial files are <code>.fastq</code> because they contain Q quality scores.</p>
<ul>
<li>there are 4 lines per each individuals sequence</li>
<li>score is series of characters-&gt;numbers-&gt;CAPletters-&gt;lowercaseletters lowest-&gt;highest</li>
<li>Q represents the quality score of nucleotides generated by automated DNA seq
<ul>
<li><code>Q = -10log10 P</code> where P = base calling error probability</li>
<li>50 would be 99.999%; 20 would be 99%; 10 would be 90%</li>
<li>Most common is <code>+33</code></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div id="downloadsymlink" class="section level2">
<h2>Download/Symlink</h2>
<p><em>All sequencing files come as zipped .fastq files</em></p>
<ol style="list-style-type: decimal">
<li><p>Make a new directory inside your project folder: <code>mkdir SOMMXXX</code></p></li>
<li>Files are always downloaded here: <code>~millermr/UCDavis/SOMM_/*index_*</code>
<ul>
<li>to learn plate barcode, open info.txt in SOMM folder <code>SOMM____ INDEX___</code></li>
<li>you can either copy the files over using <code>cp ~millermr/UCDavis/SOMM_/* ./</code></li>
<li>OR, you can make a symbolic link, e.g., : <code>ln -s /home/millermr/UCDavis/SOMM163/*.gz /home/rapeek/projects/rana_rapture/SOMM163b/raw</code> (<em>it does require full path</em>)</li>
</ul></li>
<li>To unzip files, use <code>sbatch</code> and <code>gunzip -c</code> and re-direct the output to a file of the same name (minus <code>.gz</code>):
<ul>
<li><code>sbatch -t 8:00:00 -p high gunzip -c SOMM163_S1_L001_R1_001.fastq.gz &gt; SOMM163_S1_L001_R1_001.fastq</code></li>
</ul></li>
<li><p>Once completed, move the unzipped files into your “split” folder. If you <code>cp</code> the entire <code>fastq.gz</code> files, you can remove them.</p></li>
</ol>
<div id="split-by-plate" class="section level3">
<h3>Split by Plate</h3>
<ol style="list-style-type: decimal">
<li>This takes the R2 reads, which is the plate codes, and matches to the different reads. Once done, we can delete the R2 reads and use the R1/R3 reads for the Split by Well Barcodes. This requires the following scripts:
<ul>
<li><code>run_split.sh</code></li>
<li><code>BarcodeSplitList3Files.pl</code></li>
</ul></li>
<li>Run script using something like:
<ul>
<li><code>sbatch -t 8:00:00 run_split.sh SOMM163_S1_L001_R1_001.fastq SOMM163_S1_L001_R2_001.fastq SOMM163_S1_L001_R3_001.fastq SOMM163</code></li>
</ul></li>
<li>If other plates with no data, or not your data, need to sort/filter. One option is to sort by size <code>du -hs *fastq | sort -hs</code>, get just the plates that have data…<em>(should see a size jump between plates w data and those without)</em>.
<ul>
<li>This lists all files of interest, but may need to pipe to head or tail to filter:</li>
</ul></li>
</ol>
<ul>
<li><code>du -hs *fastq | sort -hs | cut -f2</code></li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li>Move them to a folder of interest, here, the <code>split_out</code> folder in the <code>split</code> folder.</li>
</ol>
<ul>
<li><code>du -hs *fastq | sort -hs | tail -n37 | head -n-3 | cut -f2 | sed &quot;s/^/mv /g&quot; | sed &quot;s/$/ split_out\//g&quot;</code></li>
</ul>
</div>
<div id="split-by-well-barcodes" class="section level3">
<h3>Split by Well Barcodes</h3>
<ol style="list-style-type: decimal">
<li>Now we have each plate barcode, but we need to split samples by well. This requires the following scripts:
<ul>
<li><code>BarcodeSplitListBestRadPairedEnd.pl</code></li>
<li><code>run_wellsplit.sh</code> (<em>make sure executable: <code>chmod u+x script</code></em>)</li>
<li><code>run_BestRadSplit_.sh</code> (underscore is either a 6 or an 8 - depending on enzyme used)</li>
</ul></li>
<li>Make a list of the files we want to split out (need the R1/R3’s):
<ul>
<li><code>ls *R1*.fastq &gt; listR1</code></li>
<li><code>ls *R3*.fastq &gt; listR3</code></li>
</ul></li>
<li>Then cut and pipe with <code>sed</code> to add the third column (the individual plate codes for the script):
<ul>
<li><code>ls S*R1* | sed &quot;s/_R1//g&quot; | sed &quot;s/\.fastq//g&quot; | paste listR1 listR3 - &gt; flist</code></li>
</ul></li>
<li>Then we can use a <em><code>screen</code></em> to run the script to split out wells.
<ul>
<li><code>srun -t 12:00:00 -p high run_wellsplit.sh flist</code></li>
<li><code>smap -c | grep USERNAME</code></li>
<li>This runs the perl script <code>BarcodeSplitListBestRadPairedEnd.pl</code> and looks at R1 and R2 to find which has the Indiv. barcode (because the probe sits down randomly on either), then puts all the fragments with barcodes as RA and those without as RB. This should give you the same # of lines for RA and RB.</li>
<li>Script bases matches on location in file (as sorted) but a numeric value is available</li>
<li>RA will be slightly smaller (use <code>du -hs</code>) because the barcode is cleaved (but still in the name of the file)</li>
<li>this takes a good 30 minutes or more with lots of samples/plates</li>
<li>cp the files over to the <code>fastq</code> folder in your dir</li>
</ul></li>
</ol>
</div>
</div>
<div id="add-metadatasitenames-optional" class="section level2">
<h2>Add Metadata/SiteNames (Optional)</h2>
<p>If you want to add metadata to your filenames, you can do that here, but it is optional. A metadata file should contains the name of each individual, its location on the plate, etc. <em><span style="color:red"> This part can be done now or later. But be absolutely sure that your indiv names, their plate location (A1, B1, etc.), and the appropriate barcode are all in sync. For example, the barcodes might be A1 A2 A3 (Across the plate row) while your sample names are A1 B1 C1 (Down the plate column)</span></em></p>
<ol style="list-style-type: decimal">
<li>Replace barcode with name of Indiv using Metadata file and script <code>Barcode_to_Name</code>
<ul>
<li><code>sbatch -t 2880 -p high Barcode_to_Name metadatafile</code> 2880 is the number of minutes</li>
</ul></li>
<li>Concatenate each individual’s files into one RA, RB per indiv (if multiple).
<ul>
<li><code>sbatch -t 2880 -p high Concatenate.sh</code> You might need to hardcode the metadata file into this script.</li>
</ul></li>
<li>Keep concatenated files, remove all others. They should be something like <code>NAME1_RA.fastq</code> and <code>NAME1_RB.fastq</code></li>
</ol>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
